symfit/core/fit.py

```diff
@@ -1476,6 +1506,59 @@ def _init_constraints(self, constraints, model):
                     )
         return con_models
 
+    @staticmethod
+    def build_grad_lagrangian(objective, multipliers, constraint_objectives):
+        """
+        Build a :class:`~symfit.core.fit.CallableNumericalModel` representing
+        the gradient of the Lagrangian function.
+
+        The Lagrange function is
+
+        .. math:: \\mathcal{L}(\\vec{p}, \\vec{\\lambda}) = f(\\vec{p}) + \\sum_{j} \\lambda_j g_{j}(\\vec{p})
+
+        where :math:`f(\\vec{p})` is the objective function to be minimized as
+        a function of parameters :math:`\\vec{p}`, the :math:`\\lambda_j` are
+        the lagrange multipliers, and :math:`g_{j}(\\vec{p})` are the equality
+        constraints.
+
+        Because the constraints are met at stationary points of the Lagrange
+        function, we solve :math:`\\nabla_p \\mathcal{L} = 0` instead of
+        minimizing :math:`\\mathcal{L}` directly.
+
+        :param objective: Objective :math:`f(\\vec{p})`.
+        :param multipliers: list of :class:`symfit.core.argument.Parameter`
+            objects, to be interpreted as Lagrange multipliers :math:`\\lambda_j`.
+        :param constraint_objectives: list of objectives representing the
+            constraints :math:`g_{j}(\\vec{p})`.
+        :return: :class:`~symfit.core.fit.CallableNumericalModel` representing
+            :math:`\\nabla_p \\mathcal{L}`
+        """
+        grad_L = sympy.Dummy('grad_L')
+        grad_f = sympy.Dummy('grad_f')
+        if multipliers:
+            grad_gs = [constraint.model.dependent_vars[0] for constraint in
+                       constraint_objectives]
+        else:
+            grad_gs = []
+
+        # Lagrangian itself
+        model_dict = {grad_f: objective.eval_jacobian}
+        model_dict.update(
+            {grad_g_i: constraint_objectives[i].eval_jacobian for i, grad_g_i in
+             enumerate(grad_gs)})
+        model_dict[grad_L] = grad_f + sum(
+            l_i * grad_g_i for l_i, grad_g_i in zip(multipliers, grad_gs))
+
+        connectivity_mapping = {grad_f: set(objective.model.free_params)}
+        connectivity_mapping.update(
+            {grad_g_i: set(objective.model.free_params) for grad_g_i in
+             grad_gs})
+        connectivity_mapping.update({grad_L: {grad_f, *grad_gs}})
+
+        grad_lagrangian = CallableNumericalModel(model_dict,
```

Comment:
Why numerical? Don't you (also) have  an analytical form of the lagrangian?

Reply:
This is a key issue I'm having with this PR: this is indeed defined best for problems which have an analytical form. However, I'm encountering two problems with this. Apart from the obvious problem that not every model is analytical, the bigger issue is that something like our LeastSquared solver is not analytical, but the lagrange multiplier should be added to this LeastSquare, not to the model. That's a serious problem.

One solution that would keep this code very clean might be to say that this is only valid for cases where we need to `MinimizeModel` instead, since there adding the multipliers is trivial. But that would mean that at least for now it will not work for least squares fitting with data, unless the user writes out the least squares function symbolically, which is completely fair game and would work. But I don't want to do that out of the box because we cannot know how many dimension we might have to sum over etc.

Reply:
I think for a first implementation it would be acceptable to limit it to `MinimizeModel`. We'll need to think about what to do with the other Objectives.